storage/reads/merge.go

```diff
@@ -114,10 +115,47 @@ func (h *resultSetHeap) Push(x interface{}) {
 }
 
 func (h *resultSetHeap) Pop() interface{} {
-	old := h.items
-	n := len(old)
-	item := old[n-1]
-	old[n-1] = nil
-	h.items = old[0 : n-1]
+	n := len(h.items)
+	item := h.items[n-1]
+	h.items[n-1] = nil
+	h.items = h.items[:n-1]
 	return item
 }
+
+type MergedStringIterator struct {
+	iterators    []storage.StringIterator
+	uniqueValues map[string]struct{}
+	nextValue    string
+}
+
+func NewMergedStringIterator(iterators []storage.StringIterator) *MergedStringIterator {
+	return &MergedStringIterator{
+		iterators:    iterators,
+		uniqueValues: make(map[string]struct{}),
+	}
+}
+
+func (mr *MergedStringIterator) Next() bool {
+	for len(mr.iterators) > 0 {
+		iterator := mr.iterators[0]
+
+		for iterator.Next() {
+			mr.nextValue = iterator.Value()
+			if _, found := mr.uniqueValues[mr.nextValue]; !found {
```

Comment:
A `StringIterator` returned by the engine's `TagValues` or `TagKeys` APIs will produce sorted results, so you can rely on that to implement a k-way merge sort. 

Given there is still a lot to do, we can add that as a TODO in this code and come back to it later this week with a separate PR.

Reply:
I tried to find that written down somewhere, because I thought you mentioned it verbally. Anyway, #13440